arrow-libs/core/arrow-autoclose/src/commonMain/kotlin/arrow/AutoCloseScope.kt

```diff
@@ -63,50 +64,50 @@ import kotlin.coroutines.cancellation.CancellationException
  */
 public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A {
   val scope = DefaultAutoCloseScope()
+  var throwable: Throwable? = null
   return try {
     block(scope)
-      .also { scope.close(null) }
-  } catch (e: CancellationException) {
-    scope.close(e) ?: throw e
   } catch (e: Throwable) {
-    scope.close(e.throwIfFatal()) ?: throw e
+    throwable = e
+    throw e
+  } finally {
+    if (throwable !is CancellationException) throwable?.throwIfFatal()
+    scope.close(throwable)
   }
 }
 
 public interface AutoCloseScope {
+  public fun onClose(release: (Throwable?) -> Unit)
```

Comment:
What is the benefit of exposing this additional function through this interface?

Reply:
The point was to have the same basic API as ResourceScope: `onRelease` adds a handler, so `onClose` adds a handler too. This also lets `AutoCloseScope` have a SAM

Reply:
Understood. But `AutoCloseScope` is not marked as `fun interface`, right?

Reply:
No, we can mark it as such, but it's unlikely to be of benefit. The point of having it as a SAM is more of an API surface perspective: if you understand `onClose`, you understand how the whole DSL works.